bridge-history-api/internal/controller/fetcher/l1_fetcher.go

```diff
@@ -0,0 +1,261 @@
+package fetcher
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum"
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/ethclient"
+	"github.com/scroll-tech/go-ethereum/log"
+	"gorm.io/gorm"
+
+	backendabi "scroll-tech/bridge-history-api/abi"
+	"scroll-tech/bridge-history-api/internal/config"
+	"scroll-tech/bridge-history-api/internal/logic"
+	"scroll-tech/bridge-history-api/internal/orm"
+	"scroll-tech/bridge-history-api/internal/utils"
+)
+
+// L1MessageFetcher fetches cross message events from L1 and saves them to database.
+type L1MessageFetcher struct {
+	ctx             context.Context
+	cfg             *config.LayerConfig
+	db              *gorm.DB
+	crossMessageOrm *orm.CrossMessage
+	batchEventOrm   *orm.BatchEvent
+	client          *ethclient.Client
+	addressList     []common.Address
+}
+
+// NewL1MessageFetcher creates a new L1MessageFetcher instance.
+func NewL1MessageFetcher(ctx context.Context, cfg *config.LayerConfig, db *gorm.DB, client *ethclient.Client) (*L1MessageFetcher, error) {
+	addressList := []common.Address{
+		common.HexToAddress(cfg.ETHGatewayAddr),
+
+		common.HexToAddress(cfg.StandardERC20Gateway),
+		common.HexToAddress(cfg.CustomERC20GatewayAddr),
+		common.HexToAddress(cfg.WETHGatewayAddr),
+		common.HexToAddress(cfg.DAIGatewayAddr),
+
+		common.HexToAddress(cfg.ERC721GatewayAddr),
+		common.HexToAddress(cfg.ERC1155GatewayAddr),
+
+		common.HexToAddress(cfg.MessengerAddr),
+
+		common.HexToAddress(cfg.ScrollChainAddr),
+
+		common.HexToAddress(cfg.MessageQueueAddr),
+	}
+
+	// Optional erc20 gateways.
+	if cfg.USDCGatewayAddr != "" {
+		addressList = append(addressList, common.HexToAddress(cfg.USDCGatewayAddr))
+	}
+
+	if cfg.LIDOGatewayAddr != "" {
+		addressList = append(addressList, common.HexToAddress(cfg.LIDOGatewayAddr))
+	}
+
+	return &L1MessageFetcher{
+		ctx:             ctx,
+		cfg:             cfg,
+		db:              db,
+		crossMessageOrm: orm.NewCrossMessage(db),
+		batchEventOrm:   orm.NewBatchEvent(db),
+		client:          client,
+		addressList:     addressList,
+	}, nil
+}
+
+// Start starts the L1 message fetching process.
+func (c *L1MessageFetcher) Start() {
+	tick := time.NewTicker(time.Duration(c.cfg.BlockTime) * time.Second)
+	go func() {
+		for {
+			select {
+			case <-c.ctx.Done():
+				tick.Stop()
+				return
+			case <-tick.C:
```

Comment:
I think need change it to order loop. if the last fetchAndSaveEvents don't finish, the new fetchAndSaveEvents  may start.